api/src/Service/PreRegistrationVerificationService.php

```diff
@@ -97,27 +80,116 @@ public function isMultiDeputyCase(string $caseNumber): bool
     /**
      * @return PreRegistration[]
      */
-    private function applyPostcodeFilter(mixed $preRegistrationMatches, string $deputyPostcode): array
+    private function getCaseNumberMatches(array $detailsToMatchOn): array
+    {
+        /** @var PreRegistration[] $caseNumberMatches */
+        $caseNumberMatches = $this->preRegistrationRepository->findByCaseNumber($detailsToMatchOn['caseNumber']);
+
+        if (0 === count($caseNumberMatches)) {
+            $errorJson = json_encode([
+                'search_terms' => $detailsToMatchOn,
+            ]);
+
+            throw new RuntimeException($errorJson, 460);
+        }
+
+        return $caseNumberMatches;
+    }
+
+    /**
+     * @param PreRegistration[] $caseNumberMatches
+     *
+     * @return PreRegistration[]
+     */
+    private function checkOtherDetailsMatch(array $caseNumberMatches, $detailsToMatchOn)
     {
-        $deputyPostcode = DataNormaliser::normalisePostcode($deputyPostcode);
+        $matchingErrors = ['client_lastname' => false, 'deputy_lastname' => false, 'deputy_postcode' => false];
+
+        /** @var PreRegistration[] $clientLastnameMatches */
+        $clientLastnameMatches = [];
+
+        foreach ($caseNumberMatches as $match) {
+            if (mb_strtolower($match->getClientLastname()) == mb_strtolower($detailsToMatchOn['clientLastname'])) {
+                $clientLastnameMatches[] = $match;
+            }
+        }
+
+        if (0 === count($clientLastnameMatches)) {
+            $matchingErrors['client_lastname'] = true;
+            $clientLastnameMatches = $caseNumberMatches;
+        }
 
-        $preRegistrationByPostcode = [];
-        $preRegistrationWithPostcodeCount = 0;
-        foreach ($preRegistrationMatches as $match) {
-            $postcode = DataNormaliser::normalisePostcode($match->getDeputyPostCode());
+        /** @var PreRegistration[] $deputyLastnameMatches */
+        $deputyLastnameMatches = [];
 
-            if (!empty($match->getDeputyPostCode())) {
-                $preRegistrationByPostcode[$postcode][] = $match;
-                ++$preRegistrationWithPostcodeCount;
+        foreach ($clientLastnameMatches as $match) {
+            if (mb_strtolower($match->getDeputySurname()) == mb_strtolower($detailsToMatchOn['deputyLastname'])) {
+                $deputyLastnameMatches[] = $match;
             }
         }
 
-        if ($preRegistrationWithPostcodeCount < count($preRegistrationMatches)) {
-            $filteredResults = $preRegistrationMatches;
+        if (0 === count($deputyLastnameMatches)) {
+            $matchingErrors['deputy_lastname'] = true;
+            $deputyLastnameMatches = $clientLastnameMatches;
```

Comment:
I may be misunderstanding the intent here, but do we need to keep updating and creating these match variables in each block or could you just have s single variable called something like `$matches`?

Reply:
Yeah this part is slightly confusing, and unfortunately I couldn't think of a better way of writing this code.

Since we want to break down the matching process into smaller steps so we can report errors, we need to sure after each step we are getting some feedback. We also need to ensure that when there is a failure to match on a specific step, we carry on the next step, so we can record all the errors.

So put simply, at each step we are checking if a criteria matches against our array of PreRegistration records. If there are any matches, we put this into a new array.

If there isn't anything in the array, we know there wasn't any matches, and we record the error.
When there isn't a match, the next step still needs to run, but we can't pass it the empty array, so we have to pass it the last set of successful matches.

So that's why we are keeping a record of all the past matches.

Reply:
Ah OK cool - lets go with this then.